### PR TITLE
Pass non-free buffer to the MAC layer when TX_ACK is terminated.

### PR DESCRIPTION
**Pull request by hubertmis**

### CODE REVIEW

File: src/nrf_802154.h [hunk 1]
```diff
@@ -469,14 +472,15 @@ extern void nrf_802154_received(uint8_t * p_data, uint8_t length, int8_t power,
...
+ *       platform/timer directory.
+ * @note Including the timestamp for received frames uses resources like CPU time and memory. If the
+ *       timestamp is not required, use @ref nrf_802154_received_raw instead.
  *
  * @param[in]  p_data  Pointer to the buffer containing received data (PHR + PSDU). First byte in
- *                     the buffer is length of the frame (PHR). The following bytes contain the frame
- *                     itself (PSDU). The length byte (PHR) includes FCS. FCS is already verified by
+ *                     the buffer is length of the frame (PHR). The following bytes contain the
+ *                     frame itself (PSDU). The length byte (PHR) includes FCS. FCS is already
+ *                     verified by
```
Comment (b-gent) [inline src/nrf_802154.h]: can we merge lines 483 and 484 or will that be too long?

File: src/nrf_802154.h [hunk 2]
```diff
@@ -713,24 +719,60 @@ extern void nrf_802154_cca_failed(nrf_802154_cca_error_t error);
  * @brief Notify the driver that the buffer containing the received frame is not used anymore.
  *
  * @note The buffer pointed to by @p p_data may be modified by this function.
+ * @note This function can be safely called only from main context. To free buffer from callback or
```
Comment (b-gent) [inline src/nrf_802154.h]: This function can be safely called only from the main context. To free the buffer from a callback or
+ *       IRQ context, use @ref nrf_802154_buffer_free_immediately_raw.

File: src/nrf_802154.h [hunk 3]
```diff
@@ -713,24 +719,60 @@ extern void nrf_802154_cca_failed(nrf_802154_cca_error_t error);
...
+ */
+bool nrf_802154_buffer_free_immediately_raw(uint8_t * p_data);
+
 #else // NRF_802154_USE_RAW_API
 
 /**
  * @brief Notify the driver that the buffer containing the received frame is not used anymore.
  *
  * @note The buffer pointed to by @p p_data may be modified by this function.
+ * @note This function can be safely called only from main context. To free buffer from callback or
```
Comment (b-gent) [inline src/nrf_802154.h]: This function can be safely called only from the main context. To free the buffer from a callback or
IRQ context, use @ref nrf_802154_buffer_free_immediately.

File: src/nrf_802154.h (hunk 2 above)
Comment (hubertmis) [reply src/nrf_802154.h]: Fixed also in line 727

File: src/nrf_802154.h [hunk 4]
```diff
@@ -713,24 +719,60 @@ extern void nrf_802154_cca_failed(nrf_802154_cca_error_t error);
...
+ * @param[in]  p_data  A pointer to the buffer containing the received data that is no longer needed
+ *                     by the higher layer.
  */
 void nrf_802154_buffer_free_raw(uint8_t * p_data);
 
+/**
+ * @brief Notify the driver that the buffer containing the received frame is not used anymore.
+ *
+ * @note The buffer pointed to by @p p_data may be modified by this function.
+ * @note This function can be safely called from any context. If the driver is busy with processing
```
Comment (b-gent) [inline src/nrf_802154.h]: If the driver is busy processing a request called from a context with lower priority, (...)

File: src/nrf_802154.h [hunk 5]
```diff
@@ -713,24 +719,60 @@ extern void nrf_802154_cca_failed(nrf_802154_cca_error_t error);
...
+ * @param[in]  p_data  A pointer to the buffer containing the received data that is no longer needed
+ *                     by the higher layer.
  */
 void nrf_802154_buffer_free(uint8_t * p_data);
 
+/**
+ * @brief Notify the driver that the buffer containing the received frame is not used anymore.
+ *
+ * @note The buffer pointed to by @p p_data may be modified by this function.
+ * @note This function can be safely called from any context. If the driver is busy with processing
```
Comment (b-gent) [inline src/nrf_802154.h]: If the driver is busy processing a request called from a context with lower priority, (...)

**Pull request merged**